---
title: "CPDseqer Quality Control Report (multi-sample)"
date: "`r format(Sys.time())`"
output:
  html_document:
    toc: true
    toc_depth: 1
    number_sections: true
---
<style type="text/css">
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>
---

```{r setup, include=FALSE}
option_file="GSE119249.multiQC.options.txt"
knitr::opts_chunk$set(echo = TRUE,fig.width=12,fig.height=6)
library(knitr) # kable
library(reshape2) # dcast
source('Rfunctions.R')
```
```{r stat, echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
opts <- read.delim(option_file,head=F,as.is=T)
optAssignments <- paste0(opts[,1],'=\'',opts[,2],'\'')
logic.ix <- which(toupper(opts[,2])%in%c('T','F','TRUE','FALSE','NULL')) # treating assignment values T, F, NULL specially
if (length(logic.ix)>0)
  optAssignments[logic.ix] <- paste0(opts[logic.ix,1],'=',opts[logic.ix,2])
exp <- 'Experiment' # or a study
DINUC4 <- c('TT','TC','CC','CT')
cntType <- 'rCnt'
nDigit <- 3
gn <- 'hg38'
libSizeNorm <- NULL
attach(QCrefRange[[gn]])
for (assignment in optAssignments) {
  eval(parse(text=assignment))
}
if (dinuc=='DINUC4') dinuc=DINUC4
if (!exists('smps')) {
	stop('Multi-samples must be specified!')
} else {
	meta <- read.delim(smps,head=F,as.is=T)
	K <- nrow(meta)
	libSizeNorm.str <- ifelse(is.null(libSizeNorm),'NULL',libSizeNorm)
	feedback=paste0('**Study**: *',exp,'*; **Genome**: *',gn,'*; **',K,'** *samples;* **CPM_TMMnormalized**: *',libSizeNorm.str,';* **cntType**: *',cntType,'*') #; **DINUC**: *',paste0(dinuc,collapse=','),'*')
	feedback <- gsub('rCnt','ReadCount',feedback)
	feedback <- gsub('sCnt','SiteCount',feedback)
	effRange <- switch(cntType,
		rCnt=EffR.r,
		sCnt=EffR.s
	)
  contRange <- switch(cntType,
    rCnt=ContR.r,
    sCnt=ContR.s
  )
	effs <- conts <- syms <- numeric(K) # cont for contrast, different from count (file)
	counts <- meta[,2] # 2nd col has COUNT files
	bgzs <- meta[,3] # 3rd col has BGZ files
	names(effs) <- names(conts) <- names(syms) <- names(counts) <- names(bgzs) <- samples <- meta[,1]
	bedS <- vector('list',K)
	names(bedS) <- samples
	for (sample in samples) {
		#cat(sample,'\n')
		effs[sample] <- efficiency(counts[sample],cntType,gn)
		conts[sample] <- contrast(counts[sample],cntType)
		symRes <- symmetry(bgzs[sample],cntType)
		bed <- symRes$bed
		bedS[[sample]] <- bed #bed[bed[,4]%in%dinuc,]
		sym <- symRes$sym7[,dinuc]
		syms[sample] <- signif(sym[1]/sym[2],nDigit)
	}
	tbl3indices <- cbind(Efficiency=signif(effs,nDigit),Contrast=signif(conts,nDigit),Symmetry=paste(syms,1,sep=':'))
	colnames(tbl3indices)[3] <- 'Symmetry (Forward:Reverse)'
	rownames(tbl3indices) <- samples
	chrDinucMat <- compileMat_chrDinuc(counts,samples,cntType)
}
```

# Parameters for `r exp`

<font size="3"> `r feedback`</font>

* **Genome** designates the reference genome compatible with the sample. Choices are human (hg38 or hg19) and yeast (sacCer3).

* **CPM-TMMnormalized** recalls the normalization factor file used for library size normalization. The file is typically generated by *size_factor* in Step 9, which means Trimmed Mean of M-values is applied for normalization. Count-Per-Million (CPM) values below have been subjected to library size normalization. 

* **cntType** is short for Count Type, designating either Read Count or Site Count. Efficiency, Contrast, Symmetry, and chromosome-dinucleotide data matrix underlying PCA, Heatmap, and Correlation plot are all based on the specified *cntType* only.


# QC figures
<font size="4">
__Efficiency&Contrast__
</font>

**Efficiency** is defined as the ratio of read count proportion for CPD-associated dinucleotides (i.e., TT, TC, CC, and CT) over the baseline proportion based on dinucleotide composition within the reference genome (Eq. 1 in manuscript). This measure indicates at how much efficiency the CPD-Seq experiment enriches read counts on the four relevant dipyrimidine classes relative to the baseline genome composition. According to the currently accumulated public CPD-Seq datasets, Efficiency of human samples is between 2 and 3, and that of yeast samples is between 1.4 and 2.4.

**Contrast** is defined as the ratio of reads mapped to TT dinucleotides over reads mapped to AA dinucleotides, considering both forward strand and reverse strand (Eq. 2 in manuscript). In the baseline, static scenario of the reference double-stranded genome, the ratio between total TT and AA dinucleotides should be exactly 1.0 due to their complementarity. Expected Contrast range of a CPD-Seq sample (human or yeast) is 7.5~23.2.

<font size="4">
__Read Count Distribution__
</font>

The first barplot displays how many dipyrimidine sites (i.e., TT, TC, CC, and CT combined altogether), in absolute number, in the genome have the interrogated quantity of read counts. The second barplot displays how many dipyrimidine sites have the interrogated quantity of CPM values. If the user has supplied valid normalization factors through the file designated with *CPM-TMMnormalized*, the library size has been normalized to effective library size in deriving CPM values. 

<font size="4">
__Principal Component Analysis of Chromosome-Dinucleotide Counts__
</font>

Total read count (or site count, depending on *cntType*) number is summarized for each chromosome by dinucleotide class combination, and such summed read count numbers of each sample are collated to form a chromosome-dinucleotide data matrix. This data matrix forms the basis of the three plots that inform on multi-sample relationship: PCA plot, Heatmap with clustering dendrogram, and sample pairwise correlation plot. 

<font size="4">
__Heatmap of Chromosome-Dinucleotide Counts__
</font>

Hierarchical clustering uses average clustering with 1-PCC as distance, where PCC stands for Pearson Correlation Coefficient. Average linkage is used as the clustering strategy.

<font size="4">
__Sample Pairwise Correlation based on Chromosome-Dinucleotide Counts__
</font>

Pearson Correlation Coefficient is used as the correlation coefficient of a sample pair. This composite figure is rendered via *pairs.panels* from R package psych.
 
```{r figure,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
plotM_EffCont(effs,conts,effRange,contRange,exp)
knitr::opts_chunk$set(echo = TRUE,fig.width=12,fig.height=6)
plotM_rcDistrib.res <- plotM_rcDistrib(bedS,exp,gn,libSizeNorm,dinuc,pts=c(0,5,10))
detach(QCrefRange[[gn]])
write.csv(chrDinucMat,paste0(exp,'.chrDinucMat.csv')) # Will save a CSV file for chrDinuc Matrix, identified with study name.
plotPCA(t(chrDinucMat),colnames(chrDinucMat),title=paste(exp,'PCA'))
plotHeatmap(t(chrDinucMat),colnames(chrDinucMat),title=paste(exp,'Heatmap'))
plotCorrelation(chrDinucMat,title=paste(exp,'Sample Correlation'))
```

# QC tables

## Efficiency&Contrast

```{r tableEffCont,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
kable(tbl3indices)
```

## Read Count Distribution

```{r tableRcDistrib,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
kable(plotM_rcDistrib.res$tbl)
```
