---
title: "CPDseqer Quality Control Report"
date: "`r format(Sys.time())`"
output:
  html_document:
    toc: true
    toc_depth: 1
    number_sections: true
---
<style type="text/css">
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>
---

```{r setup, include=FALSE}
option_file="QC.options.txt"
knitr::opts_chunk$set(echo = TRUE,fig.width=12,fig.height=6)
library(knitr) # kable
library(reshape2) # dcast
source('Rfunctions.R')
```
```{r stat, echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
opts <- read.delim(option_file,head=F,as.is=T)
optAssignments <- paste0(opts[,1],'=\'',opts[,2],'\'')
sample <- 'Some Sample'
cntType <- 'rCnt'
gn <- 'hg38'
nDigit <- 3
attach(QCrefRange[[gn]])
for (assignment in optAssignments) {
  eval(parse(text=assignment))
}
if (!exists('bgz') | !exists('cnt')) {
	stop('Two primary input: BGZ & CNT files are not specified!')
} else {
	feedback=paste0('**Genome**: *',gn,'*; **BGZ**: *',bgz,'*; **CNT**: *',cnt,'*; **cntType**: *',cntType,'*')
	feedback <- gsub('rCnt','ReadCount',feedback)
	feedback <- gsub('sCnt','SiteCount',feedback)
	eff.r <- efficiency(cnt,'rCnt',gn)
	eff.s <- efficiency(cnt,'sCnt',gn)
	cont.r <- contrast(cnt,'rCnt')
	cont.s <- contrast(cnt,'sCnt')
	eff2 <- c(eff.r,eff.s)
	cont2 <- c(cont.r,cont.s)
	# NOTE: EffR.r, EffR.s, ContR.r, ContR.s have been defined in Rfunctions.R
	EffR.range <- paste0('(',signif(EffR.r[1],3),',',signif(EffR.r[2],3),')') # reference Range of Efficiency (rCnt)
  EffS.range <- paste0('(',signif(EffR.s[1],3),',',signif(EffR.s[2],3),')') # reference Range of Efficiency (sCnt)
	ContR.range <- paste0('(',signif(ContR.r[1],3),',',signif(ContR.r[2],3),')') # reference Range of Contrast (rCnt)
	ContS.range <- paste0('(',signif(ContR.s[1],3),',',signif(ContR.s[2],3),')') # reference Range of Contrast (sCnt)
	symRes <- symmetry(bgz,cntType)
	bed <- symRes$bed
	sym7 <- symRes$sym7
}

```
# Parameters for `r sample`

<font size="3"> `r feedback`</font>

* **Genome** designates the reference genome compatible with the sample. Choices are human (hg38 or hg19) and yeast (sacCer3).

* **BGZ** designates the dinucleotide BED file (generated in Step 6).

* **CNT** designates the dinucleotide Count file (generated in Step 6).

* **cntType** is short for Count Type, designating either Read Count or Site Count. The Symmetry plot below is based on the specified *cntType* only.

# QC figures

<font size="4">
__Efficiency&Contrast__
</font>

**Efficiency** is defined as the ratio of read count proportion for CPD-associated dinucleotides (i.e., TT, TC, CC, and CT) over the baseline proportion based on dinucleotide composition within the reference genome (Eq. 1 in manuscript). This measure indicates at how much efficiency the CPD-Seq experiment enriches read counts on the four relevant dipyrimidine classes relative to the baseline genome composition. According to the currently accumulated public CPD-Seq datasets, Efficiency of human samples is between 2 and 3, and that of yeast samples is between 1.4 and 2.4.

**Contrast** is defined as the ratio of reads mapped to TT dinucleotides over reads mapped to AA dinucleotides, considering both forward strand and reverse strand (Eq. 2 in manuscript). In the baseline, static scenario of the reference double-stranded genome, the ratio between total TT and AA dinucleotides should be exactly 1.0 due to their complementarity. Expected Contrast range of a CPD-Seq sample (human or yeast) is 7.5~23.2.

<font size="4">
__CPD-forming dinucleotides__
</font>

The first barplot displays how many dipyrimidine sites, in absolute number, in the genome have the interrogated quantity of read counts. The second barplot displays how much of each dipyrimidine class, in percentage, from the whole genome have the interrogated quantity of read counts. Such percentage values have normalized away the innate quantities of total dinucleotide sites for each dipyrimidine class, and thus are comparable across dipyrimidine classes, samples, or even species.

<font size="4">
__Forward/Reverse Strand Symmetry__
</font>

**Symmetry** measures reflects how a QC metrics is balanced between the forward strand and the reverse strand of the chromosomes. QC metrics are calculated for the forward and the reverse strands, respectively, and each pair of QC quantities are scaled to a sum of one. The scaled QC metrics are visualized in a stacked barplot overlaid with a central line that reflects the perfect symmetry of 0.5:0.5. Caution is raised if the distribution of paired QC metrics deviates obviously from the central line.

```{r figure,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
#layout(matrix(1:3,nr=1))
plot_EffCont(eff2,cont2,sample,EffR.r,EffR.s,ContR.r,ContR.s)
rcDistrib.tbl <- plot_rcDistrib(bed,sample,pts=c(0,5,10))
cat('\n\n')
plot_sym7(sym7,sample)
```

```{r tablePrepare,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
# Eff&Cont (2) tables
Eff.tbl <- cbind(round(eff2,nDigit),c(EffR.range,EffS.range))
rownames(Eff.tbl) <- c('by_ReadCount','by_SiteCount')
colnames(Eff.tbl) <- c(paste(sample,'Efficiency'),'Reference range')
Cont.tbl <- cbind(round(cont2,nDigit),c(ContR.range,ContS.range))
rownames(Cont.tbl) <- c('by_ReadCount','by_SiteCount')
colnames(Cont.tbl) <- c(paste(sample,'Contrast'),'Reference range')
# Symmetry Table
sym7.ratio <- round(t(sym7)/colSums(sym7),nDigit)
sym7.ratio.str <- paste(sym7.ratio[,1],sym7.ratio[,2],sep=':')
Sym.tbl <- cbind(t(sym7),sym7.ratio.str)
colnames(Sym.tbl) <- c('Forward (+)','Reverse (-)','Ratio') 
detach(QCrefRange[[gn]])
```
# QC tables

## Efficiency&Contrast

```{r tableEff,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
kable(Eff.tbl)
cat('\n')
kable(Cont.tbl)
```

## CPD-forming dinucleotides

```{r tableRcDistrib,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
kable(rcDistrib.tbl)
```

## Forward/Reverse Strand Symmetry

```{r tableSym7,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
kable(Sym.tbl)
```

