---
title: "CPDseqer Statistical Analysis Report"
date: "`r format(Sys.time())`"
output:
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
---

<style type="text/css">
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr) # kable
library(reshape2) # dcast
source('Rfunctions.R')
```
```{r stat, echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
opts <- read.delim('stat_sce3.options.txt',head=F,as.is=T) 
optAssignments <- paste0(opts[,1],'=\'',opts[,2],'\'')
gn='hg38'
reg <- NULL
reg1 <- 'intron'
reg2 <- 'exon'
cnt5file <- 'gn_reg_cnt5.tsv'
cntType <- 'rCnt'
for (assignment in optAssignments) {
	eval(parse(text=assignment))
}

#scenario='Statistical Comparison'
if (exists('scenario')) { # Triage based on variable sceCmmd: {stat_sce1,stat_sce2,stat_sce3}
	if (scenario=='sce1') { # NOTE: later sceCmmd can drop suffix .R
		scenario='Compare one sample with reference genome'
    if (!exists('smp')) stop('Scenario 1 (comparing one sample against reference) requires one sample\'s CNT file!')
		feedback=paste(c('**Sample** **CNT**: *',smp,'*; **Genome**: *',gn,'*; **CountType**: *',cntType,'*; **Region**: *',reg,'*'),collapse='')
		res0 <- singleSample(smp,gn,reg,cntType=cntType,cnt5file=cnt5file,nDigit=3)	
	} else if (scenario=='sce2') {
		scenario='Compare two regions in a sample'
		if (!exists('smp1')|!exists('smp2')) stop('Scenario 2 (comparing two regions) requires two regions\' CNT files!')
		feedback=paste(c('**Region1_CNT**: *',smp1,'*; **Region2_CNT**: *',smp2,'*; **Genome**: *',gn,'*; **CountType**: *',cntType,'*; **Region1_reference**: *',reg1,'*; **Region2_reference**: *',reg2,'*'),collapse='')
		res0 <- twoReg(smp1,smp2,reg1,reg2,gn,cntType=cntType,cnt5file=cnt5file)
	} else if (scenario=='sce3') {
    scenario='Compare two samples'
    if (!exists('smp1')|!exists('smp2')) stop('Scenario 3 (comparing two samples) requires two samples\' CNT files!')
		feedback=paste(c('**Sample1_CNT**: *',smp1,'*; **Sample2_CNT**: *',smp2,'*; **CountType**: *',cntType,'*'),collapse='')
		res0 <- twoGrp(smp1,smp2,cntType=cntType)
	}
	feedback <- gsub('rCnt','ReadCount',feedback)
	feedback <- gsub('sCnt','SiteCount',feedback)
	if (grepl('\\*\\*$',feedback,perl=T)) feedback=gsub('\\*\\*$','',feedback,perl=T)
}
```
# Analysis Results: `r scenario`
`r feedback`

```{r table,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
res <- res0$res
if (res0$unit==1000) { 
	unit <- toupper('thousand')
} else if (res0$unit==1e6) {
	unit <- toupper('million')
} else {
	unit <- ''
}
colnames(res)[grepl('adj$',colnames(res),perl=T)] <- paste0(grep('adj$',colnames(res),perl=T,val=T),'*')
colnames(res)[colnames(res)%in%c('exp','obs')] <- paste0(colnames(res)[colnames(res)%in%c('exp','obs')],'^')
kable(res)
```

* <font size="2">*_**Adjusted** **p-value**. Because four one-vs-others comparisons were conducted simultaneously, each exact test p-value and each Chi-squared test p-value of a one-vs-others test was multipled by four to satisfy the Bonferroni multiple test adjustment._



* ^_Expected (**exp**) situation and observed (**obs**) situation._ 

	+ _For one-vs-others comparisons, in a single-sample analysis, they are the reference/observed frequency of the concerned dinucleotide; in a two-region comparison, they appear as colon-separated ratios of expected `r unit` counts or observed `r unit` counts between the two concerned regions; in a two-group comparison, they show expected/observed `r unit` counts of the dinucleotide type in question between the two groups, separated by a comma._ 

	+ _For overall comparison or summation, in a single-sample analysis, **exp** and **obs** give the ratio between all five dinucleotide types (four actual ones + others), with counts shown in `r unit` numbers; in a two-region comparison, because it is a post summation not a genuine overall test, **exp** and **obs** do not show valid values; in a two-group comparison, a two-by-five contingency table (in `r unit` counts) is revealed in the observed (**obs**) cell, and the corresponding expected (**exp**) contingency table is revealed as well._</font>


# Explanations
## Analysis is conducted in one comparison scenario
* __Compare one (group of) sample with reference genome, with or without restriction to a region__
* __Compare two regions in a (group of) sample, considering reference background__
* __Compare two (groups of) samples, with or without restriction to a region__

In all scenarios, a region can be designated as a keyword, such as 'intron' or 'exon,' or it can be specified as a vector of dinucleotide frequencies in the reference across five categories (see below).
## Two kinds of statistical tests and test directionality
For each comparison scenarios, we present results of two kinds of statistical tests: a Chi-squared test and an exact test. Whereas Chi-squared test is always implemented for every scenario and the modalities involved, the exact test varies between Binomial test and Hypergeometric test (Fisher’s exact test). The Chi-squared test always returns a two-tailed p-value, whereas the exact test returns a one-tailed p-value.

In the context of comparing one sample with the reference genome, the exact test focuses on the right-tail, or the greater side. In the context of comparing two regions or comparing two samples, where neither of the entities is regarded as a reference, the directionality of the alternative hypothesis is indicated in column "1stVS2nd." When column 1stVS2nd shows ">," entity 1 tends to show a higher UV damage than entity 2, especially on dinucleotide TT; when 1stVS2nd shows "<," entity 1 tends to show a lower UV damage than entity 2, especially on dinucleotide TT. The statistical significance (p-value) of such one-tailed assertion is indicated in column "p.exact."

## Four concerned dinucleotide types and two comparison modalities
CPD-seqer focuses on four dinucleotide types (**TT**, **TC,** **CC**, and **CT**) that show relatively stronger UV damage signals. The other 12 types of dinucleotides are combined together to form a so-called ‘others’ type. Every comparison scenario involves two comparison modalities: the one-vs-others comparisons for each of four concerned dinucleotides, and a comparison or summary for the overall context which encompasses all four concerned dinucleotide types.

