---
title: "CPDseqer Data Analysis Report"
date: "`r format(Sys.time())`"
output:
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
---

<style type="text/css">
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>
---

```{r setup, include=FALSE}
option_file="stat_sce1.options.txt"
knitr::opts_chunk$set(echo = TRUE)
library(knitr) # kable
library(reshape2) # dcast
source('Rfunctions.R')
```
```{r stat, echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
opts <- read.delim(option_file,head=F,as.is=T) 
optAssignments <- paste0(opts[,1],'=\'',opts[,2],'\'')
logic.ix <- which(toupper(opts[,2])%in%c('T','F','TRUE','FALSE','NULL')) # treating assignment values T, F, NULL specially
if (length(logic.ix)>0)
	optAssignments[logic.ix] <- paste0(opts[logic.ix,1],'=',opts[logic.ix,2]) # =T or =F
gn='hg38'
reg <- NULL
reg1 <- 'intron'
reg2 <- 'exon'
cnt5file <- 'gn_reg_cnt5.tsv'
cntType <- 'rCnt'
libSizeNorm <- NULL
for (assignment in optAssignments) {
	eval(parse(text=assignment))
}
libSizeNorm.str <- ifelse(is.null(libSizeNorm),'NULL',libSizeNorm)
#scenario='Statistical Comparison'
if (exists('scenario')) { # Triage based on variable sceCmmd: {stat_sce1,stat_sce2,stat_sce3}
	if (scenario=='sce1') {
		scenario='Compare one sample with reference genome (Scenario I)'
    if (!exists('smp')) stop('Scenario 1 (comparing one sample against reference) requires one sample\'s CNT file!')
		feedback=paste(c('**Sample**: *',smpnames,'*; **libSize_normalization**: *',libSizeNorm.str,'*; **Genome**: *',gn,'*; **CountType**: *',cntType,'*; **Region**: *',reg,'*'),collapse='')
		res0 <- singleSample(smp,gn,reg,libSizeNorm=libSizeNorm,smpnames=smpnames,cntType=cntType,cnt5file=cnt5file,nDigit=3)	
	} else if (scenario=='sce2') {
		scenario='Compare two regions in a sample (Scenario II)'
		if (!exists('smp1')|!exists('smp2')) stop('Scenario 2 (comparing two regions) requires two regions\' CNT files!')
		feedback=paste(c('**Region1_CNT**: *',smp1,'*; **Region2_CNT**: *',smp2,'*; **libSize_normalization**: *',libSizeNorm.str,'*; **Genome**: *',gn,'*; **CountType**: *',cntType,'*; **Region1_reference**: *',reg1,'*; **Region2_reference**: *',reg2,'*'),collapse='')
		res0 <- twoReg(smp1,smp2,reg1,reg2,gn,libSizeNorm=libSizeNorm,smp1names=smp1names,smp2names=smp2names,cntType=cntType,cnt5file=cnt5file)
	} else if (scenario=='sce3') {
    scenario='Compare two samples (Scenario III)'
    if (!exists('smp1')|!exists('smp2')) stop('Scenario 3 (comparing two samples) requires two samples\' CNT files!')
		feedback=paste(c('**Sample1**: *',smp1names,'*; **Sample2**: *',smp2names,'*; **libSize_normalization**: *',libSizeNorm.str,'*; **CountType**: *',cntType,'*'),collapse='')
		res0 <- twoGrp(smp1,smp2,cntType=cntType,libSizeNorm=libSizeNorm,smp1names=smp1names,smp2names=smp2names)
	}
	feedback <- gsub('rCnt','ReadCount',feedback)
	feedback <- gsub('sCnt','SiteCount',feedback)
	if (grepl('\\*\\*$',feedback,perl=T)) feedback=gsub('\\*\\*$','',feedback,perl=T)
}
```

# Analysis Parameters: `r scenario`

<font size="3"> `r feedback`</font>

* **Genome** designates the reference genome compatible with the sample. Choices are human (hg38 or hg19) and yeast (sacCer3).

* __libSize\_normalization__ recalls the normalization factor file used for library size normalization. The file is typically generated by *size_factor* in Step 9, which means Trimmed Mean of M-values is applied before Count-Per-Million (CPM) values are derived. When the normalization factor file is void, the comparisons (Scenarios I and III) are performed on raw count numbers (site count or read count); when a valid normalization factor file is designated, the the comparisons (Scenarios I and III) are performe on CPM values.

* **CountType** designates either Read Count or Site Count.

*	**Sample**, **Sample1**, **Sample2** recapitulates the names of samples involved in comparison scenarios (I and III).

*	**Region1\_CNT**, **Rigion2\_CNT** indicates conceptual dinucleotide Count files (output of *bam2dinucleotide*) corresponding to the two region sets supplied to comparison Scenario II.

*	**Region** or **Region?\_reference** reflects background frequence of dinucleotide classes in the concerned region set. 

# Analysis Result Table

_This table includes five rows and six (or seven) columns. The top four rows show one-vs-others comparison results, each focusing on a particular dipyrimidine type. The fifth row shows the overall statistical test results. The six columns from left to right are expected and observed fractions, unadjusted and adjusted p-values of one-sided exact test, unadjusted and adjusted p-values of Chi-squire test. A p-value less than 1E-15 will be displayed as 1E-15. While Scenario I returns 6 columns in the table, Scenarios II and III returns an additional 7th column to indicate the directionality of the one-tailed exact test._  

```{r table,echo=F, comment=F, warning=F, dependson=F, message=F,results='asis'}
res <- res0$res
if (res0$unit==1000) { 
	unit <- toupper('thousand')
} else if (res0$unit==1e6) {
	unit <- toupper('million')
} else {
	unit <- ''
}
colnames(res)[grepl('adj$',colnames(res),perl=T)] <- paste0(grep('adj$',colnames(res),perl=T,val=T),'*')
colnames(res)[colnames(res)%in%c('exp','obs')] <- paste0(colnames(res)[colnames(res)%in%c('exp','obs')],'^')
kable(res)
```

* <font size="2">*_**Adjusted** **p-value**. Because four one-vs-others comparisons were conducted simultaneously, each exact test p-value and each Chi-squared test p-value of a one-vs-others test was multipled by four to satisfy the Bonferroni multiple test adjustment._



* ^_Expected (**exp**) situation and observed (**obs**) situation._ 

	+ _For one-vs-others comparisons (1st-4th rows), in a single-sample analysis (Scenario I), they are the reference/observed frequency of the concerned dinucleotide; in a two-region comparison (Scenario II), they appear as colon-separated ratios of expected `r unit` counts or observed `r unit` counts between the two concerned regions; in a two-group comparison (Scenario III), they show expected/observed `r unit` counts of the dinucleotide in question between the two groups, separated by a comma._ 

	+ _For overall comparison or summation (5th row), in a single-sample analysis (Scenario I), **exp** and **obs** give the ratio between all five dinucleotide types (four actual ones + others), with counts shown in `r unit` numbers; in a two-region comparison (Scenario II), because it is a post summation not a genuine overall test, **exp** and **obs** do not show valid values; in a two-group comparison (Scenario III), a two-by-five contingency table (in `r unit` counts) is revealed in the observed (**obs**) cell, and the corresponding expected (**exp**) contingency table is revealed as well._</font>


# Explanations
## Analysis is conducted in one of three comparison scenarios
* __I. Compare one (group of) sample with reference genome, with or without restriction to a region__ _(uv\_comp\_genome, uv\_comp\_genome\_region)_

	+ Here, we compare CPD damage of one or more UV light treated samples against the reference genome. The essential questions is if the read count number of a dipyrimidine type is significantly greater than “others”, or those non-responsive dinucleotide types (those other than TT, TC, CC, and CT), while taking into account the dinucleotide type proportion in the reference genome.

* __II. Compare two region sets in a (group of) sample, considering reference background__ _(uv\_comp\_regions)_

	+ Here, we compare CPD damage distribution between two genomic region sets within one or more samples of the same group. The essential question is if there is a significant difference in the read count number of dipyrimidine type between two genomic regions of interest, while taking into account the inherent compositions of the dipyrimidine type of these two region sets in the reference genome.  

* __III. Compare two (groups of) samples, with or without restriction to a region__ _(uv\_comp\_goups,uv\_comp\_groups\_region)_

	+ Here, we compare CPD damage between two groups of samples. The essential question is if there is a significant difference in proportion of a dipyrimidine type between the two groups. A hypergeometric test (exact test) and a Chi-squared test are used for this comparison. Unlike when comparing to the reference genome, comparing CPD damage between two groups within the same region does not require the computation of background dinucleotide information.


In Scenario I and Scenario III, the test can be performed either in a genome-wide manner or within a genomic region set (as defined via a genomic region file). CPDSeqer functions with a suffix of _\_region_ indicates the comparison is restricted to a region set.

## Two kinds of statistical tests and test directionality
For each comparison scenario, we present results of two kinds of statistical tests: a Chi-squared test and an exact test. Whereas Chi-squared test is always implemented for every **scenario** and the **modalities** involved, the exact test varies between Binomial test (Scenarios I an II) and Fisher’s exact test (Scenario III). The Chi-squared test always returns a two-tailed p-value, whereas the exact test returns a one-tailed p-value.

In the context of comparing one sample with the reference genome (Scenario I), the exact test focuses on the right-tail, or the greater side. In the context of comparing two regions (Scenario II) or comparing two samples (Scenario III), where neither of the entities is regarded as a reference, the directionality of the alternative hypothesis is indicated in column "1stVS2nd." When column 1stVS2nd shows ">," sample 1 (or group 1) tends to show a higher UV damage than sample 2 (or group 2); when 1stVS2nd shows "<," sample 1 (or group 1) tends to show a lower UV damage than sample 2 (or group 2). The statistical significance (p-value) of such one-tailed assertion is indicated in column "p.exact."

## Four concerned dinucleotide types and two comparison modalities
CPD-seqer focuses on four dinucleotide types (**TT**, **TC,** **CC**, and **CT**) that show relatively stronger UV damage signals. The other 12 types of dinucleotides are combined together to form a so-called ‘others’ type. Every comparison scenario involves two comparison **modalities**: the **one-vs-others** comparisons for each of four concerned dinucleotides, and a comparison or summary for the **overall** context which encompasses all four concerned dinucleotide types. The overall modality of Scenario II differs from Scenarios I and III because it is not an actual test but a summary across the four one-vs-others comparisons; here, the Fisher's combined probability test is applied on exact test p-values, and a Chi-squared test of summed Chi-square statistics is performed with a degree of freedom of 4.  

